core/src/main/scala/kafka/controller/ControllerContext.scala

```diff
@@ -22,6 +22,45 @@ import org.apache.kafka.common.TopicPartition
 
 import scala.collection.{Map, Seq, Set, mutable}
 
+object PartitionReplicaAssignment {
+  def fromOldAndNewReplicas(oldReplicas: Seq[Int], newReplicas: Seq[Int]): PartitionReplicaAssignment = {
+    val fullReplicaSet = (newReplicas ++ oldReplicas).distinct
+    PartitionReplicaAssignment(
+      fullReplicaSet,
+      fullReplicaSet.filterNot(oldReplicas.contains(_)),
+      fullReplicaSet.filterNot(newReplicas.contains(_))
+    )
+  }
+}
+
+case class PartitionReplicaAssignment(replicas: Seq[Int], addingReplicas: Seq[Int], removingReplicas: Seq[Int]) {
+  def isBeingReassigned: Boolean = {
+    addingReplicas.nonEmpty || removingReplicas.nonEmpty
+  }
+
+  /**
+    * Returns the partition replica assignment previous to this one.
+    * It is different than this one only when the partition is undergoing reassignment
+    * Note that this will not preserve the original ordering
+    */
+  def previousAssignment: PartitionReplicaAssignment = {
+    PartitionReplicaAssignment(
+      replicas.filterNot(addingReplicas.contains(_)),
```

Comment:
Why don't we add removingReplicas too in order to fully restore the previous state?

Reply:
Any previous assignment would not be reassigning - it would not have adding/removing replicas.
We always revert the on-going reassignment when a new one is triggered.